doc/fluid/design/dist_train/dist_train_nccl2.md

```diff
@@ -0,0 +1,34 @@
+# Distributed Training with NCCL2
+
+We design a pattern that can enable training with `ParallelExecutor` and
+using NCCL2 as it's collective communication library.
+
+In `ParallelExecutor` we can use `AllReduce` or `Reduce` and `Broadcast`
+to do multi GPU training. And if we initialize NCCL2 communicators as
+ranks in a distributed environment, we can simply run the `ParallelExecutor`
+as a distributed program! The only thing that may be different than in
+the single node version is that we need to broadcast the NCCL unique ID
+to all the nodes, and initialize communicators using that ID, so NCCL2
+will know each other as ranks.
+
+To achieve this feature, we introduce a new operator: `gen_nccl_id` op,
+so we are ***not*** "bind to" running NCCL2 with MPI, we can run it in
+what ever platform you like.
+
+It have two running modes:
+
+1. Generate and broadcast mode, which should be used on trainer 0;
+1. Listen and fetch mode, which should be used on trainers other than 0.
+
+In both two modes, this op can save the NCCL ID into current scope as a
+persistable variable, Then we can insert this op at the end of
+"startup program" of fluid, so that all workers can get the same ID to
+initialize NCCL communicator objects.
+
+<img src="src/ncc2_design.png">
```

Comment:
I have some questions about this figure, as this figure shows, at the startup stage, each trainer will generate parameters on one GPU(for example GPU0) and then broadcast the generated parameters to other GPUs(for example GPU1, GPU2, GPU3 here), because each trainer has the startup Ops, right?

Reply:
That's true, and which need to be changed. I'll update the doc to a proper design and fire an issue to get it done in later versions.